### ISSUE TITLE
Calling registerProducer() multiple times without calling unregisterProducer() is confusing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7235]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#232|
|Type|defect|
|Created|2003-08-26 06:30:23Z|

<details><summary>Searchable metadata</summary>

```
trac-id__232 232
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1061879423000000 1061879423000000
changetime__1467108217692615 1467108217692615
version__ 
owner__spiv spiv
cc__glyph cc__radix cc__exarkun cc__spiv cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
registerProducer() will happily replace an existing one with
the new producer, leading to behavior where an old producer
is never resumed again.  This seems like it would only ever
be a bug.  Would an assert or a log warning make sense in
registerProducer() to ensure no one ever does this without
realizing it?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
alternatively, the old producer should get stopProducing
called. But we do need a specific behaviour to be documented
and implemented.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I'd lean towards "explict is better than implicit", i.e.
don't magically call stopProducing.  I think a
warning.warn("registerProducer called, but a producer is
already registered") would be my preference, perhaps to
become an exception in a later release.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
seeing that doing this is a bug (since the old producer will
continue to produce if it's streaming) I suggest it be an
exception from the start. Anyone disagree?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Exception sounds good to me.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Ok, that makes sense.  I'll look at making it raise an
exception, then (seeing as someone kindly, uh, "volunteered"
me :)

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
On Tuesday, August 26, 2003, at 12:50 AM, Andrew Bennetts wrote:

> perhaps to become an exception in a later release.

I'd say make it an exception now.  AFAICT this behavior is undefined as 
it is.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
/cvs/Twisted/twisted/internet/abstract.py,v  &lt;--  abstract.py
new revision: 1.40; previous revision: 1.39
/cvs/Twisted/twisted/test/test_internet.py,v  &lt;--  test_internet.py
new revision: 1.35; previous revision: 1.34

Fixed, resolving.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
woops really resolving
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub commented|
|-|-|

In changeset a1171ec06ccf30c223674a326ab63a2c7a9e66cd
```
#!CommitTicketReference repository="" revision="a1171ec06ccf30c223674a326ab63a2c7a9e66cd"
Merge pull request #232 from twisted/py3-example-xmpp-8509

Port Jabber/XMPP examples to Python 3.

Author: ralphm
Reviewer: adiroiban
Fixes: #8509
```
